### PR TITLE
Re-enable preprocessing of pre/post JS file. NFC

### PR DESCRIPTION
**Pull request by sbc100**
This change got reverted in #19006 so this time we make the preprocessing optional.

As an example of the utility of this feature we now use pre-pocessing on `threadprofiler.js` in order to make it work on wasm64.

Fixes: #21226

### CODE REVIEW

File: src/jsifier.js [hunk 1]
```diff
@@ -97,6 +97,22 @@ function getTransitiveDeps(symbol) {
   return Array.from(transitiveDeps);
 }
 
+function shouldPreprocess(fileName) {
+  return read(fileName).match(/'use preprocessor';\n/) != null;
```
Comment (kripken) [inline src/jsifier.js:101]: should we match `"` as well as `'`? That works for `use strict` so I think maybe we should.

File: test/test_other.py [hunk 2]
```diff
@@ -14434,3 +14434,20 @@ def test_uuid(self):
...
+    console.log('assertions disabled')
+    #endif
+    ''')
+    create_file('post.js', '''
+    'use preprocessor';
+    console.log({{{ POINTER_SIZE }}});
+    ''')
+    self.emcc_args += ['--pre-js', 'pre.js', '--post-js', 'post.js']
+    self.do_runf(test_file('hello_world.c'), 'assertions enabled\n4')
+    self.do_runf(test_file('hello_world.c'), 'assertions disabled\n4', emcc_args=['-sASSERTIONS=0'])
```
Comment (kripken) [inline test/test_other.py:14453]: Please test that the string is not in the output when optimizing.

Comment (sbc100) [reply test/test_other.py:14482]: Good point.   Turns out tools don't strip such strings.. so its not a good fit.

I decided to go with `#preprocess` instead.  This is handy because it will immediately cause the parsing of such as JS file to file if you ever try to run it directly.

File: test/test_other.py [hunk 3]
```diff
@@ -14463,3 +14463,21 @@ def test_uuid(self):
...
+    #endif
+    ''')
+    create_file('post.js', '''
+    #preprocess
+    console.log({{{ POINTER_SIZE }}});
+    ''')
+    self.emcc_args += ['--pre-js', 'pre.js', '--post-js', 'post.js']
+    self.do_runf(test_file('hello_world.c'), 'assertions enabled\n4', emcc_args=['-sASSERTIONS=1'])
+    self.do_runf(test_file('hello_world.c'), 'assertions disabled\n4', emcc_args=['-sASSERTIONS=0'])
+    self.assertNotContained('use preprocessor', read_file('hello_world.js'))
```
Comment (kripken) [inline test/test_other.py:14483]:
```suggestion
    self.assertNotContained('#preprocess', read_file('hello_world.js'))
```

File: test/test_other.py (hunk 2 above)
Comment (kripken) [reply test/test_other.py:14482]: Interesting idea. `#preprocess` is nice in that it looks like a preprocessor command anyhow.

A downside is that it doesn't parse as valid JS so editors break on it, but that's the same for the other preprocessing commands anyhow. Though not `{{{ }}}`, which makes me wonder if maybe there is a way to replace `#if` etc. with something that does parse as JS... but anyhow that's separate from this.

File: ChangeLog.md [hunk 4]
```diff
@@ -24,7 +24,12 @@ See docs/process.md for more on how version tagging works.
   by name and is designed to replace all existing `-sUSE_XXX` settings for 
   ports. You can use `--show-ports` to get the list of available ports that 
   can be used with this new option. (#21214)
- 
+- `--pre-js` and `--post-js` files can now opt into being run through the JS
+  preprocessor. This change was oringally landed in  #18525, but it got reverted
```
Comment (kripken) [inline ChangeLog.md:28]:
```suggestion
  preprocessor. This change was originally landed in  #18525, but it got reverted
```

File: src/jsifier.js [hunk 5]
```diff
@@ -97,6 +97,22 @@ function getTransitiveDeps(symbol) {
   return Array.from(transitiveDeps);
 }
 
+function shouldPreprocess(fileName) {
+  return read(fileName).match(/#preprocess\n/) != null;
```
Comment (kripken) [inline src/jsifier.js:101]:
```suggestion
  return read(fileName).startsWith('#preprocess');
```

File: test/test_other.py (hunk 2 above)
Comment (sbc100) [reply test/test_other.py:14482]: Yes, @brendandahl and I have talking about this a few times.

**Comment by sbc100** [comment]
I think this is ready to go now.

@dschuff WDYT?

File: src/struct_info_internal.json [hunk 6]
```diff
@@ -24,6 +24,12 @@
           "SIGCANCEL"
         ]
     },
+    {
```
Comment (dschuff) [inline src/struct_info_internal.json:27]: is this related to the preJS preprocessing?

Comment (sbc100) [reply src/struct_info_internal.json:27]: This is needed by `threadprofiler.js` which gets included as pre-js.   It makes use of this to fix it `--thread-profiler` under wasm64.  I forgot to mention that in the description.  I will add that now.

**Review by dschuff — APPROVED** [review]
